kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/utils.kt

```diff
@@ -0,0 +1,104 @@
+package space.kscience.kmath.tensors.core
+
+import space.kscience.kmath.nd.as1D
+import space.kscience.kmath.samplers.GaussianSampler
+import space.kscience.kmath.stat.RandomGenerator
+import space.kscience.kmath.structures.*
+import kotlin.math.*
+
+/**
+ * Returns a reference to [IntArray] containing all of the elements of this [Buffer] or copy the data.
+ */
+internal fun Buffer<Int>.array(): IntArray = when (this) {
+    is IntBuffer -> array
+    else -> this.toIntArray()
+}
+
+
+/**
+ * Returns a reference to [DoubleArray] containing all of the elements of this [Buffer] or copy the data.
+ */
+internal fun Buffer<Double>.array(): DoubleArray = when (this) {
+    is DoubleBuffer -> array
+    else -> this.toDoubleArray()
+}
+
+internal inline fun getRandomNormals(n: Int, seed: Long): DoubleArray {
+    val distribution = GaussianSampler(0.0, 1.0)
+    val generator = RandomGenerator.default(seed)
+    return distribution.sample(generator).nextBufferBlocking(n).toDoubleArray()
+}
+
+internal inline fun getRandomUnitVector(n: Int, seed: Long): DoubleArray {
+    val unnorm = getRandomNormals(n, seed)
+    val norm = sqrt(unnorm.map { it * it }.sum())
+    return unnorm.map { it / norm }.toDoubleArray()
+}
+
+internal inline fun minusIndexFrom(n: Int, i: Int) : Int = if (i >= 0) i else {
+    val ii = n + i
+    check(ii >= 0) {
+        "Out of bound index $i for tensor of dim $n"
+    }
+    ii
+}
+
+internal inline fun <T> BufferedTensor<T>.minusIndex(i: Int): Int =  minusIndexFrom(this.dimension, i)
```

Comment:
Inline **must** be removed from all functions that do not take lambdas as arguments. It could be returned in some very special cases with additional justification.